### PR TITLE
feat: use protocol to force all environment states to have a key

### LINKED ISSUE #33: feat: constrain all environment states to have a key attribute
### Is your feature request related to a problem? Please describe
One assumption about environment states is that they have a `key` (jax random key) attribute to manage stochasticity in the environment step function. This key is then used in wrappers such as `JumanjiToDMEnvWrapper`. 

### Describe the solution you'd like
I am not sure of the solution to go for. I have identified two possible ways: using protocols to make it explicit that states have a `key` attribute, or using some abstract State class that will have the key attribute mandatory.

What is the best way of forcing the environment's `State` to have a `key` attribute?

### PR DESCRIPTION
**Pull request by dluo96**
Closes #33.

Implemented a protocol which ensures that every `Environment`'s `State` must implement a `key` to manage stochasticity (even if the key isn't used e.g. for deterministic environments). If a `State` is defined which doesn't implement `key`, `mypy` will throw errors. 

Currently, `mypy` does not error when a `key` attribute is implemented with a different type from `jax.random.PRNGKey`.

### CODE REVIEW

File: jumanji/env.py [hunk 1]
```diff
@@ -15,14 +15,22 @@
...
 
 from jumanji import specs
 from jumanji.types import Action, TimeStep
 
-State = TypeVar("State")
+
+class StateProtocol(Protocol):
+    """Enforce that the State for every Environment must implement a key."""
+
+    key: jax.random.PRNGKey
```
Comment (clement-bonnet) [inline jumanji/env.py:30]: I think we agreed to use chex for typing
```suggestion
    key: chex.PRNGKey
```

File: jumanji/environments/combinatorial/cvrp/types.py [hunk 2]
```diff
@@ -42,6 +44,7 @@ class State:
     visited_mask: Array  # (problem_size + 1,)
     order: Array  # (2 * problem_size,) - this size is worst-case (visit depot after each node)
     num_total_visits: jnp.int32
+    key: jax.random.PRNGKey = jax.random.PRNGKey(0)
```
Comment (clement-bonnet) [inline jumanji/environments/combinatorial/cvrp/types.py:47]: Same here
```suggestion
    key: chex.PRNGKey = jax.random.PRNGKey(0)
```

File: jumanji/environments/combinatorial/knapsack/types.py [hunk 3]
```diff
@@ -40,6 +42,7 @@ class State:
     used_mask: Array  # (problem_size,)
     num_steps: jnp.int32
     remaining_budget: jnp.float32
+    key: jax.random.PRNGKey = jax.random.PRNGKey(0)
```
Comment (clement-bonnet) [inline jumanji/environments/combinatorial/knapsack/types.py:45]:
```suggestion
    key: chex.PRNGKey = jax.random.PRNGKey(0)
```

File: jumanji/environments/combinatorial/tsp/types.py [hunk 4]
```diff
@@ -38,6 +40,7 @@ class State:
     visited_mask: Array  # (problem_size,)
     order: Array  # (problem_size,)
     num_visited: jnp.int32
+    key: jax.random.PRNGKey = jax.random.PRNGKey(0)
```
Comment (clement-bonnet) [inline jumanji/environments/combinatorial/tsp/types.py:43]:
```suggestion
    key: chex.PRNGKey = jax.random.PRNGKey(0)
```

File: jumanji/environments/games/connect4/types.py [hunk 5]
```diff
@@ -30,6 +31,7 @@
 class State:
     current_player: jnp.int8
     board: Board
+    key: jax.random.PRNGKey = jax.random.PRNGKey(0)
```
Comment (clement-bonnet) [inline jumanji/environments/games/connect4/types.py:34]:
```suggestion
    key: chex.PRNGKey = jax.random.PRNGKey(0)
```

**Review by clement-bonnet — APPROVED** [review]
Great contribution!

File: jumanji/env.py [hunk 6]
```diff
@@ -15,14 +15,22 @@
 """Abstract environment class"""
 
 import abc
-from typing import Any, Generic, Tuple, TypeVar
+from typing import Any, Generic, Protocol, Tuple, TypeVar
 
+import chex
 from chex import PRNGKey
```
Comment (clement-bonnet) [inline jumanji/env.py:21]: NIT: In this case, can we remove this line and use `chex.PRNGKey` in this module? It's more explicit.

Comment (dluo96) [reply jumanji/env.py]: Done.

**Review by clement-bonnet — APPROVED** [review]
Great stuff

**Pull request merged**